modules/src/applications/transfer/context.rs

```diff
@@ -0,0 +1,342 @@
+use crate::applications::transfer::events::{AckEvent, AckStatusEvent, RecvEvent, TimeoutEvent};
+use crate::applications::transfer::packet::PacketData;
+use crate::applications::transfer::relay::on_ack_packet::process_ack_packet;
+use crate::applications::transfer::relay::on_recv_packet::process_recv_packet;
+use crate::applications::transfer::relay::on_timeout_packet::process_timeout_packet;
+use crate::applications::transfer::{DenomTrace, HashedDenom, IbcCoin, VERSION};
+use crate::core::ics04_channel::channel::{Counterparty, Order};
+use crate::core::ics04_channel::context::{ChannelKeeper, ChannelReader};
+use crate::core::ics04_channel::msgs::acknowledgement::Acknowledgement as GenericAcknowledgement;
+use crate::core::ics04_channel::packet::Packet;
+use crate::core::ics04_channel::Version;
+use crate::core::ics05_port::context::PortReader;
+use crate::core::ics24_host::identifier::{ChannelId, ConnectionId, PortId};
+use crate::core::ics26_routing::context::{ModuleOutputBuilder, OnRecvPacketAck};
+use crate::prelude::*;
+use crate::signer::Signer;
+
+pub trait Ics20Keeper:
+    ChannelKeeper + BankKeeper<AccountId = <Self as Ics20Keeper>::AccountId>
+{
+    type AccountId;
+
+    /// Sets a new {trace hash -> denom trace} pair to the store.
+    fn set_denom_trace(&mut self, denom_trace: &DenomTrace) -> Result<(), Ics20Error>;
+}
+
+pub trait Ics20Reader: ChannelReader + PortReader {
+    type AccountId: TryFrom<Signer>;
+
+    /// get_port returns the portID for the transfer module.
+    fn get_port(&self) -> Result<PortId, Ics20Error>;
+
+    /// Returns the escrow account id for a port and channel combination
+    fn get_channel_escrow_address(
+        &self,
+        port_id: &PortId,
+        channel_id: ChannelId,
+    ) -> Result<<Self as Ics20Reader>::AccountId, Ics20Error> {
+        let hash = cosmos_adr028_escrow_address(port_id, channel_id);
+        String::from_utf8(hex::encode_upper(hash))
+            .expect("hex encoded bytes are not valid UTF8")
+            .parse::<Signer>()
+            .map_err(Ics20Error::signer)?
+            .try_into()
+            .map_err(|_| Ics20Error::parse_account_failure())
+    }
+
+    /// Returns true iff send is enabled.
+    fn is_send_enabled(&self) -> bool;
+
+    /// Returns true iff receive is enabled.
+    fn is_receive_enabled(&self) -> bool;
+
+    /// Get the denom trace associated with the specified hash in the store.
+    fn get_denom_trace(&self, denom_hash: &HashedDenom) -> Option<DenomTrace>;
+}
+
+// https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-028-public-key-addresses.md
+fn cosmos_adr028_escrow_address(port_id: &PortId, channel_id: ChannelId) -> Vec<u8> {
+    let contents = format!("{}/{}", port_id, channel_id);
+
+    let mut hasher = Sha256::new();
+    hasher.update(VERSION.as_bytes());
+    hasher.update([0]);
+    hasher.update(contents.as_bytes());
+
+    let mut hash = hasher.finalize().to_vec();
+    hash.drain(20..);
```

Comment:
Similarly here, can we use `chunks_exact()` instead of `drain()`?

Reply:
I think `truncate` is a better choice. 81a98043c9a3ec8bd2c82be6d5f4731b4a58fedc

Reply:
Absolutely!